src/Microsoft.Framework.BufferEntryCollection.Internal/BufferEntryCollection.cs

```diff
@@ -0,0 +1,199 @@
+{
+    /// <summary>
+    /// An ordered collection of strings with an enumerator that iterates over it.
+    /// </summary>
+    internal class BufferEntryCollection : IEnumerable<string>
+    {
+        // Specifies the maximum size we'll allow for direct conversion from
+        // char arrays to string.
+        private const int MaxCharToStringLength = 1024;
+        private readonly List<object> _buffer = new List<object>();
+        private bool _isModified = false;
+
+        // internal for testing
+        internal IReadOnlyList<object> BufferEntries
+        {
+            get { return _buffer; }
+        }
+
+        /// <summary>
+        /// Gets a value whether the buffer entries were modified (Added to / Cleared) or not.
+        /// </summary>
+        public bool IsModified
+        {
+            get { return _isModified; }
+        }
+
+        /// <summary>
+        /// Adds a string value to the buffer.
+        /// </summary>
+        /// <param name="value">The value to be added.</param>
+        public void Add(string value)
+        {
+            _buffer.Add(value);
+            _isModified = true;
+        }
+
+        /// <summary>
+        /// Adds a array of characters to the buffer.
+        /// </summary>
+        /// <param name="value">The array to add.</param>
+        /// <param name="index">The character position in the array at which to start copying data.</param>
+        /// <param name="count">The number of characters to copy.</param>
+        public void Add([NotNull] char[] value, int index, int count)
+        {
+            if (index < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+            if (count < 0 || value.Length - index < count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            while (count > 0)
+            {
+                // Split large char arrays into 1KB strings.
+                var currentCount = count;
+                if (MaxCharToStringLength < currentCount)
+                {
+                    currentCount = MaxCharToStringLength;
+                }
+
+                Add(new string(value, index, currentCount));
+                index += currentCount;
+                count -= currentCount;
+            }
+            _isModified = true;
+        }
+
+        /// <summary>
+        /// Adds an instance of <see cref="BufferEntryCollection"/> to the buffer.
+        /// </summary>
+        /// <param name="buffer">The buffer collection to add.</param>
+        public void Add([NotNull] BufferEntryCollection buffer)
+        {
+            _buffer.Add(buffer.BufferEntries);
+            _isModified = true;
+        }
+
+        /// <summary>
+        /// Removes all the elements from the buffer.
+        /// </summary>
+        public void Clear()
+        {
+            _buffer.Clear();
+            _isModified = true;
```

Comment:
In following scenario:
1. create a collection;
2. clear it;
<Error> `IsModify == true` even though the collection is never changed.